src/client/Client.cc

```diff
@@ -5061,11 +5062,18 @@ void Client::handle_cap_grant(MetaSession *session, Inode *in, Cap *cap, MClient
   check_cap_issue(in, cap, new_caps);
 
   // update caps
-  if (old_caps & ~new_caps) { 
-    ldout(cct, 10) << "  revocation of " << ccap_string(~new_caps & old_caps) << dendl;
+  int revoked = old_caps & ~new_caps;
+  if (revoked) {
+    ldout(cct, 10) << "  revocation of " << ccap_string(revoked) << dendl;
     cap->issued = new_caps;
     cap->implemented |= new_caps;
 
+    // break delegations if we're losing caps necessary for them
+    if (revoked & ceph_caps_for_mode(CEPH_FILE_MODE_RD))
+      in->recall_deleg(false);
```

Comment:
Whether to skip read delegations. We don't want to break read delegations on a read-only open, but for pretty much anything else we do need to recall them.